// cursor-pre-push-review/dist/git.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BASELINE_CANDIDATES = void 0;
exports.execGit = execGit;
exports.safeExecGit = safeExecGit;
exports.normalizeRemoteBaseline = normalizeRemoteBaseline;
exports.remoteRefExists = remoteRefExists;
exports.resolveEffectiveBaseline = resolveEffectiveBaseline;
exports.getMergeBase = getMergeBase;
exports.buildGitContext = buildGitContext;
exports.parsePositiveInt = parsePositiveInt;
const child_process_1 = require("child_process");
function execGit(repoRoot, args) {
    return (0, child_process_1.execFileSync)("git", args, {
        cwd: repoRoot,
        encoding: "utf8",
        maxBuffer: 10 * 1024 * 1024,
    }).trim();
}
function safeExecGit(repoRoot, args, fallback) {
    try {
        return execGit(repoRoot, args);
    }
    catch {
        return fallback;
    }
}
/** 自动选择基线时的优先级（用户未指定或指定分支不存在时） */
exports.BASELINE_CANDIDATES = [
    "origin/stable",
    "origin/dev",
    "origin/main",
    "origin/master",
];
function normalizeRemoteBaseline(branch) {
    const trimmed = branch.trim();
    if (!trimmed || trimmed === "auto")
        return "";
    return trimmed.startsWith("origin/") ? trimmed : `origin/${trimmed}`;
}
function remoteRefExists(repoRoot, ref) {
    try {
        execGit(repoRoot, ["rev-parse", "--verify", `${ref}^{commit}`]);
        return true;
    }
    catch {
        return false;
    }
}
/**
 * 解析有效基线：优先使用配置/环境变量；不存在则按 stable → dev → main → master 回退。
 */
function resolveEffectiveBaseline(repoRoot, configured) {
    const tried = [];
    const ordered = [];
    const normalized = normalizeRemoteBaseline(configured ?? "auto");
    if (normalized) {
        ordered.push(normalized);
    }
    for (const c of exports.BASELINE_CANDIDATES) {
        if (!ordered.includes(c))
            ordered.push(c);
    }
    for (const ref of ordered) {
        tried.push(ref);
        if (remoteRefExists(repoRoot, ref)) {
            return { baseline: ref, tried };
        }
    }
    return { baseline: null, tried };
}
function getMergeBase(repoRoot, baseline = "origin/main") {
    try {
        return execGit(repoRoot, ["merge-base", "HEAD", baseline]);
    }
    catch {
        return null;
    }
}
function buildGitContext(repoRoot, baseline = "origin/main") {
    const mergeBase = getMergeBase(repoRoot, baseline);
    if (!mergeBase) {
        return {
            ok: false,
            text: "(Could not compute merge-base; skip diff context.)",
        };
    }
    const logText = safeExecGit(repoRoot, ["log", "--oneline", `${mergeBase}..HEAD`], "(empty)");
    const statText = safeExecGit(repoRoot, ["diff", "--stat", `${mergeBase}..HEAD`], "(empty)");
    const diffText = safeExecGit(repoRoot, ["diff", `${mergeBase}..HEAD`], "(empty)");
    const maxDiff = parseInt(process.env.CURSOR_PRE_PUSH_MAX_DIFF_CHARS || "120000", 10);
    const truncatedDiff = diffText.length > maxDiff
        ? diffText.slice(0, maxDiff) + `\n\n[... truncated, original ${diffText.length} chars]`
        : diffText;
    const isEmpty = (!logText || logText === "(empty)") &&
        (!statText || statText === "(empty)") &&
        (!diffText || diffText === "(empty)");
    return {
        ok: true,
        mergeBase,
        isEmpty,
        text: [
            "### Review baseline (vs stable)",
            `Range **merge-base(HEAD, ${baseline})..HEAD** = commits and file changes on the current branch relative to **${baseline}**.`,
            "",
            "### merge-base",
            mergeBase,
            "",
            "### git log --oneline merge-base..HEAD",
            logText || "(empty)",
            "",
            "### git diff --stat merge-base..HEAD",
            statText || "(empty)",
            "",
            "### git diff merge-base..HEAD",
            truncatedDiff,
        ].join("\n"),
    };
}
function parsePositiveInt(raw, fallback) {
    const n = Number(raw);
    return Number.isFinite(n) && n > 0 ? Math.floor(n) : fallback;
}
